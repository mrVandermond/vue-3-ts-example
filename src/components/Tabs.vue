<template>
  <div class="tabs-container">
    <div
      v-for="tab in tabs"
      :key="tab.text"
      :class="{
        active: tab.isActive,
      }"
      class="tab"
      @click="onClickTab(tab)"
    >
      {{ tab.text }}
    </div>
  </div>
</template>

<script lang="ts" setup>
import { watchEffect, onMounted, ref } from 'vue';
import { useRoute, useRouter } from 'vue-router';
import type { ITab } from '@/types';

const tabs = ref<ITab[]>([
  {
    text: 'Каталог',
    routeName: 'catalog',
    isActive: false,
  },
  {
    text: 'Избранное',
    routeName: 'favorites',
    isActive: false,
  },
]);
const route = useRoute();
const router = useRouter();

const resetTabs = (): void => {
  tabs.value = tabs.value.map((tab) => ({
    ...tab,
    isActive: false,
  }));
};
const setActiveTab = (): void => {
  const matchedTabIndex = tabs.value.findIndex((tab) => tab.routeName === route.name);

  if (matchedTabIndex === -1) return;

  resetTabs();
  tabs.value[matchedTabIndex].isActive = true;
};

onMounted(() => {
  setActiveTab();
});

watchEffect(() => {
  setActiveTab();
});

const onClickTab = async (tab: ITab): Promise<void> => {
  await router.push({ name: tab.routeName });
};
</script>

<style lang="scss">
.tabs-container {
  display: flex;
  width: 100%;
}
.tab {
  padding: 16px 0;
  width: 372px;
  text-align: center;
  font-size: 18px;
  border-radius: 16px;
  cursor: pointer;

  &.active,
  &:hover {
    color: $cyan;
    background-color: $cyan_light;
  }
}
</style>
