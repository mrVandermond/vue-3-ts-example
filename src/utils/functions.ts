export function getItemIndexBy<T>(array: T[], item: T, field: keyof T): number {
  return array.findIndex((temp) => temp[field] === item[field]);
}
