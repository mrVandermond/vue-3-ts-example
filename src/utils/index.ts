export * as utlFuncs from './functions';
export * as utlCst from './constants';
