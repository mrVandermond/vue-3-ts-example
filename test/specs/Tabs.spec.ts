import Tabs from '@/components/Tabs.vue';

import type { TWrapperFactoryReturnType } from '../utils';
import { wrapperFactory } from '../utils';

let mockNameRoute = 'test';
const mockPushRouter = jest.fn();

jest.mock('vue-router', () => ({
  useRoute: jest.fn(() => ({
    name: mockNameRoute,
  })),
  useRouter: jest.fn(() => ({
    push: mockPushRouter,
  })),
}));

describe('Tabs component', () => {
  const getWrapper = (): TWrapperFactoryReturnType => wrapperFactory(Tabs);

  it('Рендеринг', () => {
    const wrapper = getWrapper();

    expect(wrapper.html()).toMatchSnapshot();

    const tabs = wrapper.findAll<HTMLDivElement>('.tab');

    expect(tabs.map((item) => item.classes()).join(',')).not.toContain('active');
  });

  it('Рендеринг на одном из активных табов', () => {
    mockNameRoute = 'catalog';

    const wrapper = getWrapper();
    const tabs = wrapper.findAll<HTMLDivElement>('.tab');

    expect(tabs.map((item) => item.classes()).join(',')).toContain('active');
  });

  it('Проверка обработки клика по табу', async () => {
    const wrapper = getWrapper();
    const tab = wrapper.find<HTMLDivElement>('.tab');

    await tab.trigger('click');

    expect(mockPushRouter).toHaveBeenCalledWith({ name: 'catalog' });
  });
});
