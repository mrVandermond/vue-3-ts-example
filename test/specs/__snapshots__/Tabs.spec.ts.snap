// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component Рендеринг 1`] = `
<div class="tabs-container">
  <div class="tab">
    Каталог
  </div>
  <div class="tab">
    Избранное
  </div>
</div>
`;
